bindings/python/unicorn/unicorn_py3/arch/arm.py

```diff
@@ -40,46 +40,71 @@ class UcAArch32(Uc):
     """Unicorn subclass for ARM architecture.
     """
 
+    REG_RANGE_CP = (const.UC_ARM_REG_CP_REG,)
+
     REG_RANGE_Q = range(const.UC_ARM_REG_Q0, const.UC_ARM_REG_Q15 + 1)
 
-    @staticmethod
-    def __select_reg_class(reg_id: int):
-        """Select class for special architectural registers.
+    @classmethod
+    def _select_reg_class(cls, reg_id: int) -> Type:
+        """Select the appropriate class for the specified architectural register.
         """
 
         reg_class = (
-            (UcAArch32.REG_RANGE_Q, UcReg128),
+            (UcAArch32.REG_RANGE_CP, UcRegCP),
+            (UcAArch32.REG_RANGE_Q, UcReg128)
         )
 
-        return next((cls for rng, cls in reg_class if reg_id in rng), None)
+        return next((c for rng, c in reg_class if reg_id in rng), cls._DEFAULT_REGTYPE)
 
-    def reg_read(self, reg_id: int, aux: Any = None):
-        # select register class for special cases
-        reg_cls = UcAArch32.__select_reg_class(reg_id)
+    # to learn more about accessing aarch32 coprocessor registers, refer to:
+    # https://developer.arm.com/documentation/ddi0601/latest/AArch32-Registers
 
-        if reg_cls is None:
-            if reg_id == const.UC_ARM_REG_CP_REG:
-                return self._reg_read(reg_id, UcRegCP, *aux)
+    def cpr_read(self, coproc: int, opc1: int, crn: int, crm: int, opc2: int, el: int, is_64: bool) -> int:
+        """Read a coprocessor register value.
 
-            else:
-                # fallback to default reading method
-                return super().reg_read(reg_id, aux)
+        Args:
+            coproc  : coprocessor to access, value varies between 0 and 15
+            opc1    : opcode 1, value varies between 0 and 7
+            crn     : coprocessor register to access (CRn), value varies between 0 and 15
+            crm     : additional coprocessor register to access (CRm), value varies between 0 and 15
+            opc2    : opcode 2, value varies between 0 and 7
+            el      : the exception level the coprocessor register belongs to, value varies between 0 and 3
+            is_64   : indicates whether this is a 64-bit register
 
-        return self._reg_read(reg_id, reg_cls)
+        Returns: value of coprocessor register
+        """
 
-    def reg_write(self, reg_id: int, value) -> None:
-        # select register class for special cases
-        reg_cls = UcAArch32.__select_reg_class(reg_id)
+        assert check_maxbits(coproc, 4)
```

Comment:
Why not make `cpr_read/write` return `uc_err` and avoid `assert` here by returning `UC_ERR_ARG`? Terminating the whole process seems too aggressive.

Reply:
Generally speaking, exceptions are thrown when the arguments normally make sense but there is a specific problem this time (e.g. opening a file which happens to be inaccessible, etc.). In these cases the user may catch the exception and choose how to handle it.

Here the assertions come to block an invalid usage with incorrect arguments: it is not a call that may sometimes work and sometimes not -- rather it is invalid. There is no way to "handle" such case other than re-writing that piece of code.

Reply:
Okay, I also forget `AssertionError` can be caught so let it go.